cdap-watchdog/src/main/java/co/cask/cdap/metrics/process/DataMigrator.java

```diff
@@ -0,0 +1,177 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Migrates data from v2 table to v3 table with salting.
+ */
+public class DataMigrator {
+  private static final Logger LOG = LoggerFactory.getLogger(DataMigrator.class);
+
+  private final LinkedHashMap<Integer, Integer> maxRecordsToScanResolutionMap;
+  private final ScheduledExecutorService scheduledExecutorService;
+  private final MigrationTableUtility migrationTableUtility;
+  private final CConfiguration cConf;
+
+
+
+  /**
+   * Data migrator to schedule data migration for metrics tables from v2 to v3 metrics tables.
+   * @param metricDatasetFactory metric dataset factory
+   * @param cConf cConfiguration
+   * @param hConf hConfiguratioin
+   * @param hBaseTableUtil Hbase table util
+   * @param maxRecordsToScanResolutionMap linked hash map of resolutions to perform migration
+   *                                      to the max records to scan for that resolution table
+   *                                      while performing migration
+   * @param executorService scheduled executor service
+   */
+  public DataMigrator(MetricDatasetFactory metricDatasetFactory, CConfiguration cConf, Configuration hConf,
+                      HBaseTableUtil hBaseTableUtil, LinkedHashMap<Integer, Integer> maxRecordsToScanResolutionMap,
+                      ScheduledExecutorService executorService) {
+    this.migrationTableUtility = new MigrationTableUtility(cConf, hConf, metricDatasetFactory, hBaseTableUtil);
+    this.maxRecordsToScanResolutionMap = maxRecordsToScanResolutionMap;
+    this.scheduledExecutorService = executorService;
+    this.cConf = cConf;
+  }
+
+  /**
+   * if the metrics data migration is already completed, we would just shutdown the executor service and return
+   * if not, we would schedule migration task at fixed rate schedule.
+   * when the migration is complete, the task would shutdown the executor service by itself
+   */
+  public void scheduleMigrationIfNecessary() {
+    if (!isMigrationComplete()) {
+      // we have 5 minute initial delay for availability of transaction/dataset service
+      // to be able to get the metrics table or if v3 tables are not already created they might take some time
+      // initially, we don't have to perform
+      scheduledExecutorService.scheduleAtFixedRate(new MigrationTask(), 5, 1, TimeUnit.MINUTES);
+      LOG.info("Scheduled metrics migration thread for resolution INT_MAX, 3600, 60 tables");
+    } else {
+      scheduledExecutorService.shutdown();
+    }
+  }
+
+  /**
+   * returns true when data migration is complete and
+   * all v2 metrics tables data has been migrated for configured resolutions ; false otherwise
+   * @return true if migration is complete; false otherwise
+   */
+  private boolean isMigrationComplete() {
+    Iterator<Integer> resolutions = maxRecordsToScanResolutionMap.keySet().iterator();
+
+    if (!resolutions.hasNext()) {
```

Comment:
this is not really needed because in this case, the while loop will not execute.